### PR TITLE
Bugfix/2537 missing close brace err msg

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Fix for issue https://github.com/stan-dev/stan/issues/2537 - parser fails without error message when program is missing final closing brace.

Stan also gives poor error messages when non-final blocks or statement sequences are missing a closing brace.  Used this bugfix to change processing of variable name so that we can detect likely cases where missing close brace causes parser to run off the rails.

The issue reported is due to the interaction between `stan::io::program_reader` and boost spirit qi parser;  the parser's line position iterator increments line number beyond actual number of lines in file.  When error occurs at end of file, the line position iterator line is greater than end line number recorded by the program_reader.  instead of reporting the error message, the parser throws an exception.  To fix this problem, we pad the end of line position by 2.  Fixed file `stan/io/program_reader.hpp` and updated unit tests accordingly.

Fix was found because helper file `stan/test/unit/lang/utililty.hpp` had function which created stub program_reader object with end line fixed at 500.  This was very useful for this issue, however, overall, this is a bad thing because unit tests which read in a test model from disk should use a proper program_reader object to read the file in from disk so that unit test behavior is as close as possible to behavior of Stan compiler.  Therefore got rid of function `create_stub_reader`.

- unit  tests which read in a test-model file from disk now instantiate a program reader and use its `read` method to record start and end line numbers.

- unit tests for where Stan programs which are coded directly into the unit test as a string (some parser tests) or the program is coded up directly as an AST with no corresponding program (some generator tests) still create a stub reader object is appropriate - updated these tests with comment to that effect.

#### Intended Effect
1. Better error messages for unclosed program blocks.
2. Unit tests process model files the same was as compiler does.

#### How to Verify
Unit tests.

#### Side Effects
Better error messages.

#### Documentation
N/A

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/stan/lang/grammars/semantic_actions_def.cpp [hunk 1]
```diff
@@ -1606,6 +1606,14 @@ namespace stan {
     boost::phoenix::function<deprecate_increment_log_prob>
     deprecate_increment_log_prob_f;
 
+    // called from: statement_grammar
```
[the remark this answers is not in the document]
Comment (mitzimorris) [reply src/stan/lang/grammars/semantic_actions_def.cpp]: made all changes as requested.

**Comment by mitzimorris** [comment]
thanks Ben!

**Comment by bbbales2** [comment]
Np, want me to look at your new pull req.?

**Comment by mitzimorris** [comment]
many thanks - I've asked Matthijs to do that one - it's all about
the spirit qi parser which is crazy.  the code that got changed
doesn't even look like c++.

if you're curious, check out file
src/stan/lang/grammars/statement_grammar_def.hpp

cheers,
Mitzi

**Pull request merged**